utxo/src/utxo_impl/mod.rs

```diff
@@ -0,0 +1,537 @@
+
+        if let Some(res) = self.utxos.get(key) {
+            return Some(res.clone());
+        }
+
+        // since the utxo does not exist in this view, try to check from parent.
+        self.parent.and_then(|parent| {
+            parent.get_utxo(outpoint).map(|utxo| UtxoEntry {
+                // if the utxo exists in parent:
+                // dirty is FALSE because this view does not have the utxo, therefore is different from parent
+                // fresh is FALSE because this view does not have the utxo but the parent has.
+                status: UtxoStatus::Entry(utxo),
+                is_dirty: false,
+                is_fresh: false,
+            })
+        })
+    }
+
+    pub fn new(parent: &'a dyn UtxosView) -> Self {
+        UtxosCache {
+            parent: Some(parent),
+            current_block_hash: parent.get_best_block_hash(),
+            utxos: BTreeMap::new(),
+            memory_usage: 0,
+        }
+    }
+
+    pub fn set_best_block(&mut self, block_hash: Id<Block>) {
+        self.current_block_hash = Some(block_hash);
+    }
+
+    pub fn add_utxos(
+        &mut self,
+        tx: &Transaction,
+        source: UtxoSource,
+        check_for_overwrite: bool,
+    ) -> Result<(), Error> {
+        let id = OutPointSourceId::from(tx.get_id());
+
+        for (idx, output) in tx.get_outputs().iter().enumerate() {
+            let outpoint = OutPoint::new(id.clone(), idx as u32);
+
+            let overwrite = if check_for_overwrite {
+                self.has_utxo(&outpoint)
+            } else {
+                // by default no overwrite allowed.
+                false
+            };
+
+            let utxo = Utxo {
+                output: output.clone(),
+                // TODO: where do we get the block reward from the transaction?
+                is_block_reward: false,
+                source: source.clone(),
+            };
+
+            self.add_utxo(utxo, &outpoint, overwrite)?;
+        }
+        Ok(())
+    }
+
+    /// Mark the inputs of tx as 'spent'.
+    /// returns a TxUndo if function is a success;
+    /// or an error if the tx's input cannot be spent.
+    pub fn spend_utxos(&mut self, tx: &Transaction, height: BlockHeight) -> Result<TxUndo, Error> {
+        let tx_undo: Result<Vec<Utxo>, Error> = tx
+            .get_inputs()
+            .iter()
+            .map(|tx_in| self.spend_utxo(tx_in.get_outpoint()))
+            .collect();
+
+        self.add_utxos(tx, UtxoSource::BlockChain(height), false)?;
+
+        tx_undo.map(TxUndo::new)
+    }
+
+    /// Adds a utxo entry in the cache.
+    pub fn add_utxo(
+        &mut self,
+        utxo: Utxo,
+        outpoint: &OutPoint,
+        possible_overwrite: bool,
+    ) -> Result<(), Error> {
+        let key = outpoint;
+        let is_fresh = match self.utxos.get(key) {
+            None => {
+                // An insert can be done. This utxo doesn't exist yet, so it's fresh.
+                !possible_overwrite
```

Comment:
Generally speaking, a utxo can be overwritten if it's spent. This is motivated in bitcoin by BIP30, where coinbases can have identical hashes. Apart from that issues, there's no harm in overwriting utxos as long they're spent, which I don't think will ever happen. I find it unnecessary because the chance to have a collision is very low, it's there in bitcoin. It's a point of discussion whether we want to drop this feature. Otherwise it may be there for reasons that I still don't comprehend.